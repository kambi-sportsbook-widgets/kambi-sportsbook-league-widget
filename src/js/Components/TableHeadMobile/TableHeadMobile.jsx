import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { translationModule } from 'kambi-widget-core-library'
import { DropdownButton } from 'kambi-widget-components'
import TableHeadDefaultRow from '../TableHeadDefaultRow/TableHeadDefaultRow'
import styles from './TableHeadMobile.scss'

class TableHeadMobile extends Component {
  constructor(props) {
    super(props)

    this.state = {
      columnGroupIdx: this.props.initialColumnGroupIdx,
    }
  }

  /**
   * Handles column group change.
   * @param {number} idx Column group key
   */
  columnGroupChanged(idx) {
    this.setState({ columnGroupIdx: idx })
    this.props.onColumnGroupChanged(idx)
  }

  /**
   * Returns currently selected columns group definition.
   * @returns {object}
   */
  get columnGroup() {
    return this.props.columnGroups[this.state.columnGroupIdx]
  }

  /**
   * Renders table header for mobile devices.
   * @returns {XML}
   */
  render() {
    const columnPickerOptions = this.props.columnGroups.map(column => {
      return translationModule.getTranslation(column.title)
    })
    let headerCssClasses = [
      'KambiWidget-card-header-border',
      'KambiWidget-secondary-header',
      'KambiWidget-card-support-text-color',
      styles['mobile-row'],
    ]
    headerCssClasses = headerCssClasses.join(' ')

    return (
      <thead>
        <tr className={headerCssClasses}>
          <th colSpan="2" className={styles.title}>
            {this.props.title}
          </th>
          <th
            colSpan={this.columnGroup.columns.length}
            className={styles['column-picker']}
          >
            {this.props.showColumnPicker ? (
              <DropdownButton
                options={columnPickerOptions}
                selected={this.props.initialColumnGroupIdx}
                onChange={this.columnGroupChanged.bind(this)}
              />
            ) : null}
          </th>
          <th className={styles.margin} />
        </tr>
        <TableHeadDefaultRow columns={this.columnGroup.columns} />
      </thead>
    )
  }
}

TableHeadMobile.propTypes = {
  /**
   * Widget's title
   */
  title: PropTypes.string.isRequired,

  /**
   * Column groups definitions
   */
  columnGroups: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      columns: PropTypes.arrayOf(
        PropTypes.shape({
          name: PropTypes.string.isRequired,
          short: PropTypes.string.isRequired,
        })
      ).isRequired,
    })
  ).isRequired,

  /**
   * If true shows the column picker
   */
  showColumnPicker: PropTypes.bool.isRequired,

  /**
   * Defines which column group should be displayed upon component creation
   */
  initialColumnGroupIdx: PropTypes.number,

  /**
   * Called on columns configuration change (with column group index argument)
   */
  onColumnGroupChanged: PropTypes.func.isRequired,
}

TableHeadMobile.defaultProps = {
  initialColumnGroupIdx: 0,
}

export default TableHeadMobile
