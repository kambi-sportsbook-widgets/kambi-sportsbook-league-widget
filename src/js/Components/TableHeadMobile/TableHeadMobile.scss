@import '../TableHeadShared';

:local(.mobile-row) {
   th {
      @extend .table-head-cell;

      &.column-picker {
         text-align: right;
      }
   }
}

:local(.mobile-row-non-collapsable) {
   @extend .table-head-row;
}
