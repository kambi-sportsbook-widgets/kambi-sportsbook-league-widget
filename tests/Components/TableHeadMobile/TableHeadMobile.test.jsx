/* eslint-env jest */
import React, { Children } from 'react';
import TableHeadMobile from '../../../src/js/Components/TableHeadMobile/TableHeadMobile';
import ReactTestRenderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

jest.mock('kambi-widget-core-library', () => ({
   translationModule: {
      getTranslation: (key) => `Translated: "${key}"`
   }
}));

const columnGroups = [
   {
      id: 'cg1',
      title: 'Column group 1',
      columns: [
         {name: 'Column 1-1', short: 'c11'},
         {name: 'Column 1-2', short: 'c12'}
      ]
   },
   {
      id: 'cg2',
      title: 'Column group 2',
      columns: [
         {name: 'Column 2-1', short: 'c21'},
         {name: 'Column 2-2', short: 'c22'}
      ]
   }
];

describe('TableHeadMobile DOM rendering', () => {

   it('renders correctly when not collapsable', () => {
      const tree = ReactTestRenderer.create(
         <TableHeadMobile
            title="Test title"
            columnGroups={columnGroups}
            onHeadClick={() => false}
            onColumnGroupChanged={() => false}
            collapsable={true}
         />
      ).toJSON();

      expect(tree).toMatchSnapshot();
   });

   it('renders correctly when collapsable', () => {
      const tree = ReactTestRenderer.create(
         <TableHeadMobile
            title="Test title"
            columnGroups={columnGroups}
            onHeadClick={() => false}
            onColumnGroupChanged={() => false}
            collapsable={false}
         />
      ).toJSON();

      expect(tree).toMatchSnapshot();
   });

   it('renders correctly when in hidden mode', () => {
      const tree = ReactTestRenderer.create(
         <TableHeadMobile
            title="Test title"
            columnGroups={columnGroups}
            onHeadClick={() => false}
            onColumnGroupChanged={() => false}
            collapsable={true}
            hiddenMode={true}
         />
      ).toJSON();

      expect(tree).toMatchSnapshot();
   });

   it('renders correctly with initial column group specified', () => {
      const tree = ReactTestRenderer.create(
         <TableHeadMobile
            title="Test title"
            columnGroups={columnGroups}
            onHeadClick={() => false}
            onColumnGroupChanged={() => false}
            collapsable={true}
            initialColumnGroupIdx={1}
         />
      ).toJSON();

      expect(tree).toMatchSnapshot();
   });

});

describe('TableHeadMobile interface', () => {

   it('handles head click events correctly', () => {
      const onHeadClickMock = jest.fn();

      const wrapper = shallow(
         <TableHeadMobile
            title="Test title"
            columnGroups={columnGroups}
            onHeadClick={onHeadClickMock}
            onColumnGroupChanged={() => false}
            collapsable={true}
            hiddenMode={true}
         />
      );

      expect(onHeadClickMock).not.toHaveBeenCalled();

      wrapper.first().simulate('click');

      expect(onHeadClickMock).toHaveBeenCalledTimes(1);
   });

   it('handles onColumnGroupChange events correctly', () => {
      const onColumnGroupChangedMock = jest.fn(),
         clickEvMock = { stopPropagation: jest.fn() };

      const wrapper = mount(
         <table>
            <TableHeadMobile
               title="Test title"
               columnGroups={columnGroups}
               onHeadClick={() => false}
               onColumnGroupChanged={onColumnGroupChangedMock}
               collapsable={true}
               hiddenMode={false}
            />
         </table>
      );

      wrapper.find('button').simulate('click', clickEvMock);

      const eventMock = new Event('click');
      Object.defineProperty(
         eventMock,
         'target',
         {
            value: wrapper.find('li[data-kw-dropdown-button-index=1]').node,
            enumerable: true
         }
      );

      expect(onColumnGroupChangedMock).not.toHaveBeenCalled();

      window.document.documentElement.dispatchEvent(eventMock);

      expect(onColumnGroupChangedMock).toHaveBeenCalledTimes(1);
      expect(onColumnGroupChangedMock).toHaveBeenCalledWith(1);
   });

});
