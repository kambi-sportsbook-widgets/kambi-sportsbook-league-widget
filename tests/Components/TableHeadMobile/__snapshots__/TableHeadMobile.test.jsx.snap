// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHeadMobile DOM rendering renders correctly 1`] = `
<thead>
  <tr
    className="KambiWidget-card-header-border KambiWidget-secondary-header KambiWidget-card-support-text-color mobile-row"
  >
    <th
      className="title"
      colSpan="2"
    >
      Test title
    </th>
    <th
      className="column-picker"
      colSpan={2}
    >
      <DropdownButton
        horizontalAlignment="right"
        onChange={[Function]}
        options={
          Array [
            "Translated: \\"Column group 1\\"",
            "Translated: \\"Column group 2\\"",
          ]
        }
        selected={0}
        verticalAlignment="top"
      />
    </th>
    <th
      className="margin"
    />
  </tr>
  <TableHeadDefaultRow
    columns={
      Array [
        Object {
          "name": "Column 1-1",
          "short": "c11",
        },
        Object {
          "name": "Column 1-2",
          "short": "c12",
        },
      ]
    }
    hiddenMode={false}
    title=""
  />
</thead>
`;

exports[`TableHeadMobile DOM rendering renders correctly when showColumnPicker is false 1`] = `
<thead>
  <tr
    className="KambiWidget-card-header-border KambiWidget-secondary-header KambiWidget-card-support-text-color mobile-row"
  >
    <th
      className="title"
      colSpan="2"
    >
      Test title
    </th>
    <th
      className="column-picker"
      colSpan={2}
    />
    <th
      className="margin"
    />
  </tr>
  <TableHeadDefaultRow
    columns={
      Array [
        Object {
          "name": "Column 1-1",
          "short": "c11",
        },
        Object {
          "name": "Column 1-2",
          "short": "c12",
        },
      ]
    }
    hiddenMode={false}
    title=""
  />
</thead>
`;

exports[`TableHeadMobile DOM rendering renders correctly with initial column group specified 1`] = `
<thead>
  <tr
    className="KambiWidget-card-header-border KambiWidget-secondary-header KambiWidget-card-support-text-color mobile-row"
  >
    <th
      className="title"
      colSpan="2"
    >
      Test title
    </th>
    <th
      className="column-picker"
      colSpan={2}
    >
      <DropdownButton
        horizontalAlignment="right"
        onChange={[Function]}
        options={
          Array [
            "Translated: \\"Column group 1\\"",
            "Translated: \\"Column group 2\\"",
          ]
        }
        selected={1}
        verticalAlignment="top"
      />
    </th>
    <th
      className="margin"
    />
  </tr>
  <TableHeadDefaultRow
    columns={
      Array [
        Object {
          "name": "Column 2-1",
          "short": "c21",
        },
        Object {
          "name": "Column 2-2",
          "short": "c22",
        },
      ]
    }
    hiddenMode={false}
    title=""
  />
</thead>
`;
